src/particles/particle_creation.hpp

```diff
@@ -135,6 +139,241 @@ template <ParticleType particleType> struct ParticleCreationTraits {
+										 state_arr(ii, jj, kk, HydroSystem<problem_t>::density_index);
+									rho_adj = state_arr(ii, jj, kk, HydroSystem<problem_t>::density_index);
+									numx += rho_adj * (vx_adj - (vx)) * (vx_adj - (vx));
+									numy += rho_adj * (vy_adj - (vy)) * (vy_adj - (vy));
+									numz += rho_adj * (vz_adj - (vz)) * (vz_adj - (vz));
+
+									denominator += rho_adj;
+								}
+							}
+						}
+						sigma_sq_x = numx / denominator;
+						sigma_sq_y = numy / denominator;
+						sigma_sq_z = numz / denominator;
+
+						p.rdata(mass_idx + 1) = (std::abs(vx) / vx) * amrex::RandomNormal(std::abs(vx), std::sqrt(sigma_sq_x), engine);
+						p.rdata(mass_idx + 2) = (std::abs(vy) / vy) * amrex::RandomNormal(std::abs(vy), std::sqrt(sigma_sq_y), engine);
+						p.rdata(mass_idx + 3) = (std::abs(vz) / vz) * amrex::RandomNormal(std::abs(vz), std::sqrt(sigma_sq_z), engine);
+
+						// Keep generating mass until it smaller than max mass in Sukhbold but smaller than high mass limit
+						double mass_of_star = 1.e50;
+						while (mass_of_star > m_imf_max || mass_of_star < m_star_high) {
+							const double xx = amrex::Random(engine);
+							mass_of_star = xx * (std::pow(m_imf_max, 1.0 - alpha) - std::pow(m_imf_min, 1.0 - alpha)) +
+								       std::pow(m_imf_min, 1.0 - alpha);
+							mass_of_star = std::pow(mass_of_star, 1. / (1. - alpha));
+							p.rdata(mass_idx) = mass_of_star;
+						}
+
+						total_momx += p.rdata(mass_idx + 1) * p.rdata(mass_idx);
+						total_momy += p.rdata(mass_idx + 2) * p.rdata(mass_idx);
+						total_momz += p.rdata(mass_idx + 3) * p.rdata(mass_idx);
+
+						p.idata(evolution_stage_index) = interpolate_fate(p.rdata(mass_idx));
+						p.rdata(birth_time_index + 1) = interpolate_death_time(p.rdata(mass_idx));
+						;
+					}
+				}
+
+				if (num_particles > 1) { // Update momentum of the low mass star if there is(are) high mass star(s) in the cell
+					const int p_idx = 0;
+					auto &plow = particles[p_idx]; // NOLINT
+					plow.rdata(mass_idx + 1) = -total_momx / plow.rdata(mass_idx);
+					plow.rdata(mass_idx + 2) = -total_momy / plow.rdata(mass_idx);
+					plow.rdata(mass_idx + 3) = -total_momz / plow.rdata(mass_idx);
+				}
+				state_arr(i, j, k, HydroSystem<problem_t>::density_index) = (cell_mass - particle_mass) / cell_volume;
+				vx = state_arr(i, j, k, HydroSystem<problem_t>::x1Momentum_index) / state_arr(i, j, k, HydroSystem<problem_t>::density_index);
+				vy = state_arr(i, j, k, HydroSystem<problem_t>::x2Momentum_index) / state_arr(i, j, k, HydroSystem<problem_t>::density_index);
+				vz = state_arr(i, j, k, HydroSystem<problem_t>::x3Momentum_index) / state_arr(i, j, k, HydroSystem<problem_t>::density_index);
+				state_arr(i, j, k, HydroSystem<problem_t>::internalEnergy_index) *=
+				    cell_density / state_arr(i, j, k, HydroSystem<problem_t>::density_index);
+				state_arr(i, j, k, HydroSystem<problem_t>::energy_index) =
```

Comment:
You don't need to separate out the velocities. The cell energy should just decrease in proportion to the mass decrease, so that you leave the energy per unit mass of the gas unchanged.